### ISSUE TITLE
Write the design and implementation document

### ISSUE THREAD

**Issue by charitarthchugh** [charitarthchugh/paperscale#31]
## Question

Not a decision — the destination. Write the single design and implementation document that
this map exists to produce, synthesized from every closed ticket.

It must be readable start to finish by someone who has not read the map, and complete enough
to implement from without re-deriving anything. Cover:

- the domain and its vocabulary (cross-reference `CONTEXT.md`)
- the adapter contract and what it is allowed to know
- chunking, pooling, and the uniform output shape
- both sinks, with their exact formats
- identity and resume, including the user-facing warning about re-OCR
- the TUI panel
- concurrency, batching, and failure handling
- the implementation plan: module layout and a commit sequence

Decide while writing where it lives — `docs/`, a top-level design file, or the README — and
whether the pinned-model rationale belongs in it or in user-facing docs.

## Blocked by

Every other ticket on this map. When this closes, the map closes. Four remain open, all raised
by an independent reconstruction of this document from the recorded decisions alone — the
reconstruction succeeded, which is why these are gaps in the *connective tissue* rather than in
any decision:

- [The Adapter's concrete contents, and the Chunk budget as numbers](https://github.com/charitarthchugh/paperscale/issues/37)
- [The four handoffs #30 never picked up](https://github.com/charitarthchugh/paperscale/issues/40) — itself blocked by #37
- [The panel's server group: who produces its rows](https://github.com/charitarthchugh/paperscale/issues/38)
- [Clear the stale text from the closed record](https://github.com/charitarthchugh/paperscale/issues/39)

A full independent draft already exists and should be read before writing this one — it is
faithful to every decision and its gaps section is the source of the four tickets above. It is
**not** a substitute for this ticket: it inherits the holes rather than filling them.

**Comment by charitarthchugh** [comment]
From #29 (resolved: variant A). The implementation plan must include three prerequisite changes to `src/paperscale/tui.py`, none of which are embed-specific bugs but all of which block the chosen panel: (1) `_stat_columns()` hardcodes group order `("run", "vllm", "issues")`, which renders embed's groups as `run | issues | server` — throughput lands last; (2) `_layout_budget()` pads the event pane with blank rows on a quiet run; (3) resumed skips must be visually distinct from work performed, since a large skip count is the only symptom of the stale-output failure that standing decision 7 deliberately leaves undetected.

**Comment by charitarthchugh** [comment]
From [The LanceDB schema and its namespacing](https://github.com/charitarthchugh/paperscale/issues/27), now closed — two things for the write-up.

**The dependency list is now fixed:** **numpy** (from #26), **lancedb** and **pyarrow** (from #27). None of them is a paperscale dependency today — all are absent from `dependencies`, from `[project.optional-dependencies]` (which holds only `tui = ["rich"]`) and from the dev group. Whether `embed` ships them as hard dependencies or as an extra following the `tui` precedent is on the map under *Not yet specified*. Everything measured for #26 and #27 was against **lancedb 0.37.1 / pyarrow 25.0.1 / numpy 2.5.2**, and two API notes belong in the plan: `create_scalar_index` is deprecated in favour of `create_index(col, config=BTree())`, and `create_index(metric=...)` in favour of `create_index(col, config=IvfPq(...))`.

**Two CLI items surfaced.** The LanceDB database path is a new argument with no prior art in this repo. And run labels are neither path-safe nor validated — `_parse_runs` at `src/paperscale/cli.py:65` only strips whitespace and checks non-empty and unique, while standing decision 6 and #22 put the label into a filesystem path, so `--run 'legal/2024=...'` silently creates a nested directory. Both are recorded on the map under *Not yet specified* alongside the rest of the CLI surface.

**Comment by charitarthchugh** [comment]
From [How resume knows a document is done](https://github.com/charitarthchugh/paperscale/issues/28), now closed — three items for the write-up.

**The manifest block is eight facts, not seven.** `layout` (`bare` | `labelled`) joins #26's invariant block, guarding the hazard #22 flagged: the run-label directory appears only when one Invocation embeds more than one Run, so a second Invocation with a different run set would write to paths that match nothing and silently duplicate the corpus into a parallel subtree.

**Name derivation changed, and it changes the OCR side too.** Standing decision 6 now reads: `os.path.normpath` → `lstrip("/")` → drop remaining `..` and empty parts. The existing sanitizer at `src/paperscale/pipeline.py:686-689` drops `..` rather than resolving it and keeps `.`, so `/a/b/../c.pdf` and `/a/b/c.pdf` derive the same name. Because embed mirrors the markdown export's rule, **the two must not diverge** — if the export keeps the old sanitizer, the side-by-side layouts standing decision 6 wants stop matching. The plan has to say which side changes and in what order.

**The standing-decision-7 user warning is drafted** and sits in [#28's resolution](https://github.com/charitarthchugh/paperscale/issues/28#issuecomment-5327197210), ready to place. It also documents the one gap normalization cannot close: `docs/a.pdf` and `/home/cc/docs/a.pdf` are irreconcilable because nothing records the OCR-time working directory.

Also worth carrying into the plan: `--no-resume` **deliberately diverges** from the OCR pipeline's `_wipe_workspace_progress`, which `rmtree`s results and flags. Embed's `--no-resume` re-embeds and overwrites and deletes nothing, because an embed output is a deliverable rather than scratch. That divergence needs saying out loud in user docs, or it reads as a bug.

**Comment by charitarthchugh** [comment]
From [Concurrency, batching, and what the panel measures](https://github.com/charitarthchugh/paperscale/issues/30), now closed — **the map's design decisions are complete.** One ticket now stands between this one and the destination: [The CLI surface: flags, defaults, and how embed installs](https://github.com/charitarthchugh/paperscale/issues/35), which graduated the last three items out of the map's *Not yet specified*.

Four things from #30 for the write-up.

**The panel reads `prompt_tps`, never `gen_tps`.** Embeddings are prefill-only, so `vllm:generation_tokens_total` never moves. This is a one-field difference from the OCR panel that is easy to miss and impossible to notice at runtime — the wrong field simply reads zero, which looks like an idle server rather than a bug. `Rates` already exposes both, so nothing needs writing; it only needs getting right once.

**Three defaults ship explicitly unmeasured** — the per-request token budget, the concurrency limit, and the request retry ceiling. No live embedding server existed to measure against, so the design marks them as starting points rather than tuned values, and names `vllm:num_requests_waiting` as the instrument for tuning concurrency. #35 decides what actually ships.

**Requests mix Documents and split on terminal failure.** Because a token budget fills a request from several Documents, one poison Document would otherwise take every Document sharing its request down with it. On terminal failure the batch is re-issued one Document at a time, and only those that fail alone are recorded failed.

**`<out>/paperscale-embed-failures.txt` joins `<out>/paperscale-embed.json`** under #26's reserved-name collision guard, and the run exits non-zero when any Document failed.

Two corrections to carry: #29's resolution listed `empty` in both the `run` and `issues` groups — it belongs in `run` only, since #28 made an empty Document a recorded outcome with a real output rather than a problem, and the freed `issues` slot shows `retrying`. And #29's three `tui.py` prerequisites are unchanged and still required before variant A renders correctly.

**Comment by charitarthchugh** [comment]
Now blocked by two tickets, not one: [The CLI surface: flags, defaults, and how embed installs](https://github.com/charitarthchugh/paperscale/issues/35) and [How comparable pipelines are actually implemented](https://github.com/charitarthchugh/paperscale/issues/36).

#36 is deliberately scheduled to run **immediately before** this ticket rather than earlier. It reads how comparable pipelines — self-hosted OpenAI-compatible server, local sinks, resume over a large corpus — are implemented in code, and sorts every decision this map made into conventional / deliberately divergent / unconventional-for-an-unstated-reason. Only the third category matters, and per #23's precedent a finding that contradicts a locked decision is raised and stopped on, never reversed inside the research ticket.

The reason for the ordering: run earlier, it becomes input to decisions instead of a check on them, and the map would be designing by imitation rather than from its own measurements.

**Comment by charitarthchugh** [comment]
From [How comparable pipelines are actually implemented](https://github.com/charitarthchugh/paperscale/issues/36), now closed — **this is the only ticket left on the map.** Five things it must carry, and one of them is unresolved.

1. **Correct #30's batching rationale.** vLLM fans an `input` array of N texts into **N independent engine requests** (verified by hand against source). Request batching amortizes HTTP round trips only; it does not help the engine batch. So concurrency, not batch size, is what fills the engine — which makes `vllm:num_requests_waiting` a more direct instrument than #30 assumed, and makes split-on-failure cheaper than #30 assumed. The decision stands; the stated reasoning does not.

2. **State that Chunk vectors are the primary artifact and the Document vector is a convenience over them.** Zero of fourteen comparable pipelines produce a document-level vector as a first-class output. Standing decision 3's reason holds, but the divergence is unanimous, so a Consumer will not expect it and must be told plainly.

3. **State that persisting vectors is a deliberate premise.** Not universal — nomic-embed keeps them in RAM, consumes them for filtering, and discards them, treating embeddings as a recomputable intermediate.

4. **Carry "read signatures, not prose" into the implementation plan.** Several targets' docstrings contradict their own code, including LanceDB's retry default (docstring says 10, signature says 7).

5. **Unresolved — the concurrency challenge.** #30 chose fixed concurrency on the reasoning that no engine publishes a ceiling to derive one from. Vespa's feed client adapts without needing one, by measuring `throughput / inflight^0.3`. This must be settled before the concurrency section is written; it is with the map owner now.

**Comment by charitarthchugh** [comment]
## Correction to my own #36 note, plus three decisions this document must carry

**Item 2 of my earlier note from [How comparable pipelines are actually implemented](https://github.com/charitarthchugh/paperscale/issues/36) was wrong and must not be written as stated.**
It said this document should *"state explicitly that Chunk vectors are the primary artifact and the
Document vector is a convenience, since no comparable pipeline produces the latter."* That advice
was drawn from a sample that turned out to be homogeneous.

### 1. Two Consumers, two primary artifacts — do not rank them

paperscale's output feeds **two** Consumers:

- **RAG** — retrieve text, hand it to a language model. Reads the **Chunk vectors**. A Document
  vector is at most a coarse-to-fine filter here, and is genuinely the weaker instrument for
  ranking, because max-over-Chunks is query-*dependent* and a stored mean cannot be.
- **Classification** — the vectors are features for a separate model predicting a per-Document
  label. Reads the **Document vectors**. `sklearn`/`xgboost`/torch want `X` of shape
  `(n_documents, stored_dim)` against one label per Document; that *is* the `document_vector`
  array. There is no query, so the reduction cannot be deferred to read time — it has to happen
  somewhere, and the only question was where.

All fourteen pipelines in #36 are retrieval pipelines. Every one ends at a search index, which is
exactly the job that never needs a Document vector. **The unanimity was an artifact of the sample,
not evidence against standing decision 3.** State it that way: the divergence is a *different job*,
not a trade we lost. Written plainly, this also stops a future reader deleting the Document vector
on the authority of that research file, which is now checked into the repo.

### 2. Both Sinks carry both artifacts; the guidance is which Sink to reach for

Splitting the artifacts across the Sinks — Chunk vectors to LanceDB for RAG, Document vectors to
`.npz` for classification — was considered and **rejected**. It costs four things:

1. It puts **coarse-to-fine RAG across a cross-Sink join**: a database query joined against a
   directory tree.
2. It leaves [#27](https://github.com/charitarthchugh/paperscale/issues/27)'s `documents` table
   without a reason to exist, and that table holds provenance `chunks` deliberately does *not*
   repeat — `source_file`, `source_digest`, `created`.
3. It spends the property [#34](https://github.com/charitarthchugh/paperscale/issues/34) and
   [#25](https://github.com/charitarthchugh/paperscale/issues/25) earned across two tickets: that
   **one** Sink alone is enough to recompute the Document vector from `chunk_vectors` +
   `token_count` and confirm it. Split them and verification needs both Sinks and a join.
4. It makes both Sinks mandatory for a Consumer doing both jobs. `CONTEXT.md` says a Run *may* have
   more than one Sink.

So [#26](https://github.com/charitarthchugh/paperscale/issues/26) and #27 are unchanged. What this
document adds is guidance, and it is **not** the split a reader might guess:

| Sink | What it is for |
|---|---|
| **LanceDB** | The working store, for **both** jobs. Vector index and filters for RAG; one columnar scan of `documents` builds the classification feature matrix; both vectors in one place, so coarse-to-fine works. |
| **`.npz`** | Portability and archive. No database, plain files, numpy the only thing needed to read it. |

The axis is *queryable store vs portable files*, not *RAG vs classification*.

**Measured, because it is the one place the guidance bites.** Building `X` from the `.npz` Sink
means opening one ZIP per Document under #26's mirrored-tree layout. Warm cache, single-chunk
layout, all eight arrays present:

```
2000 docs, 768-dim
  per-document .npz :  206.9 ms   (103 us/doc)
  single .npy matrix:    1.7 ms
  ratio: 124x
  extrapolated   100,000 docs:  10.3 s  vs 0.08 s
  extrapolated 1,000,000 docs: 103.4 s  vs 0.84 s
```

Real, but not disqualifying — seconds to a couple of minutes, once, and cacheable by the Consumer.
An Invocation-level `document_vectors.npy` roll-up was rejected on that number: it is an aggregate
written once at the end, so it fights #26's per-Document atomicity and
[#28](https://github.com/charitarthchugh/paperscale/issues/28)'s resume story — a resumed Invocation
would have to rebuild it by reading every `.npz` anyway.

Also worth stating for the classification Consumer, since it keeps #25 from looking like a
constraint: the token-weighted mean is a **default feature**, not the only one available. A Consumer
wanting first-Chunk, max-pool, or mean⊕max concat builds any of them from `chunks`, and `n_chunks`
and `token_count` are themselves usable features.

### 3. Raw vector norms — considered and declined

Every stored vector is L2-normalized, so magnitude is destroyed and #25 confirmed the raw norms are
neither stored nor recoverable. Reclaiming them is cheap in principle — request
`use_activation: false`, record `‖raw‖` as one `float32` per Chunk, then slice-and-normalize
client-side, which #34's algebra already proves is equivalent. **Declined:** for these models the
raw norm tracks token count and frequency far more than meaning, length is already available via
`token_count`, and it would put a new server-side precondition on the request path to buy a feature
no Consumer has asked for. Record it as declined rather than unexamined — a classification Consumer
is exactly who would wonder.

---

### Two items on my earlier notes now settled

- **Item 5** — *"Resolve the concurrency challenge before the concurrency section is written."*
  Settled: fixed default retained, plus a panel advisory when `vllm:num_requests_waiting` stays above
  zero for a sustained window. Advisory only, no control loop. See
  [#30's amendment](https://github.com/charitarthchugh/paperscale/issues/30#issuecomment-5335725615).
- **Standing decision 2 amended on the map.** It asserted overflow is "routine, not exceptional".
  Measured against the 49-Document smoke corpus it is **3 of 49** (~6%), stable across chars/token
  ratios of 3.0–4.0 — median Document 8,880 chars, then a cliff to 121k/171k/218k. Nothing
  downstream reverses: #24's greedy packer and #25's token weighting are right for a rare-but-large
  tail, and that tail is real (89 pages). But this document should describe Overflow as the minority
  path, because that sentence is what a reader uses to judge how much the chunking machinery is
  worth.

**Comment by charitarthchugh** [comment]
From [The Adapter's concrete contents, and the Chunk budget as numbers](https://github.com/charitarthchugh/paperscale/issues/37), now closed — three things this document inherits.

**The module-layout question this ticket was told to answer is now half-answered.** `embed` is its own package, `src/paperscale/embed/`, holding the **whole subcommand** and not only the Adapters — mirroring `src/paperscale/evaluation/`, including the `_handle_evaluate` shape where every heavy import sits inside the handler function rather than at module scope. That is what makes #35's "one `embed` extra, imports inside the handler" implementable: `src/paperscale/models/__init__.py` eagerly imports all nine OCR Adapters, so embed Adapters placed there would pull embed code into every OCR Run. **The split of files inside that package, and the commit sequence, remain this ticket's.**

**The Adapter section has its complete contract.** Five typed class attributes, one Adapter per model **size** rather than per family, and no default entry in the registry:

```python
class EmbedModel(abc.ABC):
    card_context_length: int      # the card's number; min()'d against the server at run time
    native_dim: int               # published per model; doubles as the wrong-model assertion (#34)
    min_dim: int = 32             # Qwen3's published floor, adopted as the convention
    document_instruction: str     # applied by paperscale
    query_instruction: str        # recorded for the Consumer; never applied
```

**The provenance invariant block grows.** `instruction` is now **two** facts, `document_instruction` and `query_instruction`, both invariant within an Invocation. The block moves from seven entries to eight — nine for the `.npz` Sink, which #28 also gave `layout`. Both Sinks must carry both strings, since a Consumer matching query-side conventions reads them from the Sink rather than from a model card.

Two things worth writing into the document as facts, not as background:

- **`document_instruction` is `""` for Qwen3 — empty string, never null.** `""` says *we decided on none*; a missing value says *we did not record it*. Provenance has to keep those apart, and a reader of the schema will otherwise "tidy" one into the other.
- **Omitting Qwen3's query-side Instruction costs approximately 1–5% retrieval performance**, published by Qwen. paperscale never applies it, but the Consumer decides whether to, and cannot decide without the number.
